File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int>cmap;
    int index = 0;
    TreeNode* arrayToTree(vector<int>&preorder, int left, int right){
        if(left > right) return nullptr;
        int rootvalue = preorder[index++];
        TreeNode* root_val = new TreeNode(rootvalue);
        root_val->left = arrayToTree(preorder, left, cmap[rootvalue]-1);
        root_val->right = arrayToTree(preorder, cmap[rootvalue]+1, right);
        return root_val;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for(int i = 0;i<inorder.size();i++){
            cmap[inorder[i]] = i;
        }
        return arrayToTree(preorder, 0, preorder.size()-1);
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        inorder_index_map = {val : i for i, val in enumerate(inorder)}
        self.index = 0

        def build(left, right):
            if left > right:
                return None
            
            root_val = preorder[self.index]
            self.index += 1
            root = TreeNode(root_val)
            mid = inorder_index_map[root_val]
            root.left = build(left, mid -1)
            root.right = build(mid + 1, right)

            return root
        
        return build(0, len(inorder)-1)      